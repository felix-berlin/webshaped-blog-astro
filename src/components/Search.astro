---
import "@pagefind/default-ui/css/ui.css";
---

<div class="c-search" id="search"></div>

<script>
  import { PagefindUI } from "@pagefind/default-ui";

  let bundlePath = `${import.meta.env.BASE_URL}pagefind/`;

  if (import.meta.env.DEV) {
    bundlePath = "/dist/pagefind/";
    console.info(
      "If there are no search results, make sure you have run `npm run build` and that the `dist/pagefind` folder exists.",
    );
  }

  const initPagefind = () => {
    new PagefindUI({
      element: "#search",
      resetStyles: false,
      showImages: false,
      bundlePath,
    });
  };

  window.addEventListener("DOMContentLoaded", initPagefind);

  document.addEventListener("astro:after-swap", initPagefind);

  window.addEventListener("keydown", (event) => {
    if (event.key === "/" || event.key === ".") {
      event.preventDefault();
      document?.querySelector("div#search input")?.focus();
    }
  });
</script>
