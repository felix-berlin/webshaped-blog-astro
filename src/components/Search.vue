<template>
  <div :id="id" class="c-search"></div>
  <!-- <input type="text" ref="searchInput" @focus="initPagefind" @keydown="processResults" /> -->
</template>

<script setup lang="ts">
import { onMounted, onUnmounted, nextTick } from "vue";
import { PagefindUI } from "@pagefind/default-ui";

export interface SearchProps {
  id: string;
}

const { id } = defineProps<SearchProps>();

let bundlePath = `${import.meta.env.BASE_URL}pagefind/`;

if (import.meta.env.DEV) {
  bundlePath = "/dist/pagefind/";
  console.warn(
    "If there are no search results, make sure you have run `npm run build` and that the `dist/pagefind` folder exists.",
  );
}

/**
 * Initialize Pagefind
 *
 */
const initPagefind = async () => {
  new PagefindUI({
    element: `#${id}`,
    resetStyles: false,
    showImages: false,
    bundlePath,
  });
};

const processResults = async (event) => {
  const results = await (await pagefind.value.search(event.target.value)).results;
  for (const result of results) {
    const data = await result.data();
    console.log(data, data.meta.title, data.excerpt);
    // do required DOM manipulation
  }
};

/**
 * Trigger search via keyboard
 *
 * @param   {KeyboardEvent}  event
 *
 * @return  {void}
 */
const triggerSearchViaKeyboard = (event: KeyboardEvent): void => {
  const activeElement = document.activeElement?.tagName;
  // If input or textarea is focused, do nothing
  if (activeElement === "INPUT" || activeElement === "TEXTAREA") return;

  if (!(event.key === "/" || event.key === ".")) {
    return;
  }
  event.preventDefault();
  const inputElement = document?.querySelector(`#${id} input`) as HTMLElement;
  inputElement?.focus();
};

onMounted(() => {
  nextTick(() => {
    initPagefind();
  });

  window.addEventListener("keydown", (event) => triggerSearchViaKeyboard(event));
});

onUnmounted(() => {
  window.removeEventListener("keydown", (event) => triggerSearchViaKeyboard(event));
});
</script>
<style lang="scss">
@use "@pagefind/default-ui/css/ui.css";
</style>
