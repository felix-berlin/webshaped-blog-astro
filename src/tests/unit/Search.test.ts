import { mount } from "@vue/test-utils";
import { JSDOM } from "jsdom";
import { it, expect, describe } from "vitest";
// @ts-ignore: Unresolved import
import Search from "@components/Search.vue";

describe("Search", async () => {
  it("Search component renders correctly", async () => {
    // Create a new JSDOM instance
    const dom = new JSDOM("<!DOCTYPE html><html><body></body></html>");

    // Set the global window and document objects to the JSDOM window and document objects
    global.window = dom.window;
    global.document = dom.window.document;

    // Mount the Search component
    const wrapper = mount(Search, {
      props: {
        id: "search",
      },
    });

    // Assert that the component renders correctly
    expect(wrapper.html()).toMatchSnapshot();
  });
});
