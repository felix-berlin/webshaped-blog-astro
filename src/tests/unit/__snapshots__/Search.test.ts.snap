// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > Search component renders correctly 1`] = `"<div id="search" class="c-search"></div>"`;
